### PR TITLE
Simplify recursive data structure in CFC Vindex

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR simplifies the self-reference of the CFC struct by prefix Vindex in the CFC Vindex.
Prefix Vindex is created when it is required rather than upfront. This resolved the cyclic dependency which was causing cache_size to crash with stack full error.

## Related Issue(s)

- Fixes CFCInsert Flakyness reported via launchable
- Fixes https://github.com/vitessio/vitess/issues/11858

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [X] Tests were added or are not required
-   [X] Documentation was added or is not required

### CODE REVIEW

**Comment by GuptaManan100** [comment]
The unit tests are failing. Could you take a look at those?

**Comment by harshit-gangal** [comment]
> The unit tests are failing. Could you take a look at those?

updated test expectation.

**Comment by harshit-gangal** [comment]
cc: @wangmeng99

File: go/vt/vtgate/executor_framework_test.go [hunk 1]
```diff
@@ -177,7 +177,10 @@ var executorVSchema = `
 			"params": {
 				"region_bytes": "1"
 			}
-    	}
+    	},
+		"cfc": {
+          "type": "cfc"
```
Comment (deepthi) [inline go/vt/vtgate/executor_framework_test.go:182]: The formatting looks off. I'm surprised that `gofmt` is passing though.

Comment (GuptaManan100) [reply go/vt/vtgate/executor_framework_test.go:182]: This is in a string, so I think that's why `gofmt` is not complaining

File: go/vt/vtgate/vindexes/cfc.go [hunk 2]
```diff
@@ -244,7 +241,8 @@ func (vind *CFC) Map(ctx context.Context, vcursor VCursor, ids []sqltypes.Value)
 
 // PrefixVindex switches the vindex to prefix mode
 func (vind *CFC) PrefixVindex() SingleColumn {
-	return vind.prefixVindex
+	// prefixCFC is only used in 'LIKE' compare expressions.
+	return &prefixCFC{CFC: vind}
 }
```
Comment (GuptaManan100) [inline go/vt/vtgate/vindexes/cfc.go:246]: Everytime we call this `PrefixVindex()` function, it would return a new struct. Is that what we want?

**Comment by deepthi** [comment]
Does this failure occur only in CI, or is it possible in production? If it can fail in production we should document the expected failure as a bug (with an issue).

File: go/vt/vtgate/vindexes/cfc.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vtgate/vindexes/cfc.go:246]: that is not ideal, let me re-write this part.

**Comment by harshit-gangal** [comment]
> Does this failure occur only in CI, or is it possible in production? If it can fail in production we should document the expected failure as a bug (with an issue).

It is a production bug for users using CFC vindex.

**Review by GuptaManan100 — APPROVED** [review]
Looks good to me now other than the indentation Deepthi pointed out

File: go/vt/vtgate/executor_framework_test.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vtgate/executor_framework_test.go]: fixed.

**Comment by harshit-gangal** [comment]
> Looks good to me now other than the indentation Deepthi pointed out

that is resolved.

**Review by systay — COMMENTED** [review]
LGTM. Only thing missing would an issue for the bugfix, right?

**Pull request merged**